server/aesdsocket.c

```diff
@@ -0,0 +1,342 @@
+
+  ret = bind(socket_fd, conn_info->ai_addr, conn_info->ai_addrlen);
+  if(ret < 0)
+  {
+    perror("bind");
+    return -1;
+  }
+
+  freeaddrinfo(conn_info);
+
+  //Starting the deamon
+  if(deamon)
+    {
+      pid_t pid;
+
+      pid = fork ();
+      if (pid == -1)
+      {
+          free(read_buf);
+          close(socket_fd);
+          perror("fork");
+          return -1;
+      }
+      else if (pid != 0)
+      {
+          exit (EXIT_SUCCESS);
+      }
+
+      else //Valid pid
+      {
+        if(setsid()==-1)
+        {
+            free(read_buf);
+            perror("session");
+            close(socket_fd);
+            return -1;
+        }
+        if(chdir("/")==-1)
+        {
+            free(read_buf);
+            perror("chdir");
+            close(socket_fd);
+            return -1;
+        }
+        close(STDIN_FILENO);
+        close(STDOUT_FILENO);
+        close(STDERR_FILENO);
+        open ("/dev/null", O_RDWR);
+        dup (0); 
+        dup (0); 
+      }
+    }
+
+
+  //Listening on the port
+  ret = listen(socket_fd, 5);  
+  if(ret < 0)
+  {
+    perror("listen");
+    return -1;
+  }
+
+  struct sockaddr_storage their_addr;
+  socklen_t addr_size = sizeof their_addr;
+
+
+  int  new_fd = 0;
+  f_fd = open("/var/tmp/aesdsocketdata",O_RDWR | O_CREAT | O_TRUNC | O_APPEND, 755); //666 equivalent value
+
+
+  
+
+  size_t len, total_len;
+  int r_ret = 0;
+  bool packet_complete = false;
+  uint64_t realloc_count = 1;
+
+
+  if (signal (SIGINT, signal_handler) == SIG_ERR) 
+  {
+    perror("signal()");
+    return -1;
+  }
+  if (signal (SIGTERM, signal_handler) == SIG_ERR) 
+  {
+    perror("signal()");
+    return -1;
+  }
+
+
+
+  while(!signal_received)
+  {
+    new_fd = accept(socket_fd, (struct sockaddr *)&their_addr, &addr_size);
+    if(new_fd< 0)
+    {
+      perror("accept");
+      return -1;
+    }
+
+    inet_ntop(their_addr.ss_family,get_in_addr((struct sockaddr *)&their_addr),s, sizeof s);
+    syslog(LOG_DEBUG, "Accepted connection from %s",s);
+    printf("Accepted connection from %s",s);
+
+    finished_writing = true;
+
+    while((!packet_complete))
+    {
+      
+      r_ret = recv(new_fd, (read_buf+((realloc_count-1)*1024)), 1024, 0);
+      
+      if(r_ret == 0)
+      {
+        //Connection is closed
+        printf("Accepted connection from %s",s);
+        syslog(LOG_DEBUG, "Closed connection from %s",s);
+        break;
+      }
+      else if(r_ret < 1)
+      {
+        perror("recv");
+        free(read_buf); //Freeing
+        return -1;
+      }
+
+      //len = strlen(read_buf);
+      len = ((realloc_count-1)*1024) + r_ret;   
+
+      total_len = is_newline_found(read_buf,len);
+
+      if(total_len > 0)
+        packet_complete = true;
+
+      if(!packet_complete)
+      {
+        realloc_count++;
+        read_buf = realloc(read_buf, realloc_count*1024);
+      }
+        
+      if(packet_complete)
+      {
+        write_to_file(f_fd, read_buf, len);
```

Comment:
Does this handle case where newline is followed by more characters in the same packet?